### DISCUSSION TITLE
CVRP with deliveries and pickups to and from the depot

### DISCUSSION THREAD

**Discussion by bansalaarav** [google/or-tools#3376] [Routing (and legacy CP) questions]
Hello,

I’m trying to use OR-tools to solve a CVRP with deliveries and pickups problem. Say there is a company that leases out vehicles. On any day the company has some customers to send their vehicles too and some customers to collect their vehicles from. For example: if the company has to collect 4 vehicles and deliver 3, it can send 3 vehicles (the ones that have to be delivered) with 4 drivers (2 in 1 vehicle, 1 each in the other 2). On the way back, these 4 drivers will be driving the 4 collection vehicles. The objective is to find the shortest route that the 3 vehicles can take.

Since the vehicles are going from and coming to the depot, I’m trying to just edit the CVRP OR-tools example. In my current code, I’ve set the demand for a delivery node to -1 and the demand for a collection node to +1. I’ve also shortened the distance_matrix so that there are 8 total nodes (4 collection + 3 delivery + 1 depot). I’ve set the num_vehicles to 3 and the capacity of the vehicles to [1, 2, 1]. Here the ‘capacity’ refers to the number of drivers in each vehicle. Since there can’t be 0 drivers in a vehicle, in the capacity dimension, I’ve added a lower bound using SetCumulVarSoftLowerBound.

When I run my code the solution utilises only 1 vehicle but there are 3 deliveries, so all 3 vehicles have to necessarily travel. Even when I increase the number of drivers to 6 (2 each in each vehicle) I get the same result.

How do I make it so all of the delivery vehicles move while also solving my initial problem - finding the shortest path.

Here’s the [link](https://colab.research.google.com/drive/15AUd_UUub-OEu5IbgLTBpwqoUFWi4UXa?usp=sharing) to my colab file.

**Comment by jmarca** [comment]
Honestly it is hard to figure out what you are talking about.  I suspect you are mixing up concepts, which is leading to the issues you are seeing.

Dimensions are accumulators.  You are correct in assuming you can add 1 and subtract 1 to represent pickups and deliveries.  But the idea that the capacity is a driver, and that drivers can morph into vehicles?  That has me puzzled.  I can't see any way that the solver will do what you want it to do, because it is picking up and delivering drivers, not vehicles.

I did not click on the link.  I did not look at your code.

Sorry that probably isn't very helpful.

**Comment by Mizux** [comment]
If I were you I would.
* Create a driver dimension, each "Vehicle to collect" location will consume 1 driver.
* Each vehicle as an implicit driver you didn't count aka the dimension is only use to track extra driver you need to "deliver" 
* Duplicate each "Vehicle to collect" node, the duplicate node will be the start node of a new vehicle aka you deliver a driver then you can reuse the vehicle which now have a driver in it.
* If you have a time dimension add a constraint: the time dimension's CumulVar of `n_1` < time dimension's CumulVar of `n_1_duplicate` (aka `v1_start`) so you can simulate the vehicle 1 start its route once the driver (in vehicle 0) has been delivered.

Here the idea, you have a vehicle `v_0` at depot `d` and a vehicle `v1` at node `n_1` to pickup (aka deliver a driver at `n_1`), then suppose you'll want to use this vehicle `v_1` to visit/serve node `n_2`.
So, you'll have vehicle `v_0` (starting from `depot`) which will visit `n_1` to "deliver" a driver and pickup the vehicle `v_1` then after this vehicle will perform a visit.
And vehicle v_1 is starting from `n_1_duplicate`.

Something like this:
```mermaid
flowchart LR
v0_s(v0 start) -->|"v_0 (driver: 1)"| n_1 -->|"v_0 (driver: 0)"| v0_e(v0 end)
v1_s(n_1_duplicate aka v1 start) -->|"v_1 (driver: 0)"| n_2 -->|"v_1 (driver: 0)"| v1_e(v1 end)
n_1 -..->|precedence constraint| v1_s
```

**Reply by bansalaarav** [reply]
Thank you so much for the reply! This is my first time using or-tools so I would really appreciate some code related to the implementation of the driver dimension...
